### ISSUE TITLE
0.9 RC1->Final Nil Operator Expression Regression

### ISSUE THREAD

**Issue by wied03** [opal/opal#1274]
Appears to have been introduced with https://github.com/opal/opal/commit/cbb05e53de526cd234e18a4ce5f095b9e27d129d but that is probably not the root cause of the problem.

Problem appears to be related to operator precedence and usage of parentheses with the `nil` global in JS. I've observed the problem with ternary expressions and guards.

Test Source:

```ruby
def return_nil
  %x{
    return nil;
  }
end

class Foo
  def <(other)
    %x{
      return nil;
    }
  end
end

puts "no class #{return_nil ? 'yes' : 'no'}"
puts "class w/o parens #{Foo.new < 2 ? 'yes' : 'no'}"
puts "class w/ parens #{(Foo.new < 2) ? 'yes' : 'no'}"
puts "guard w/ parens" unless (Foo.new < 2)
puts "guard w/o parens" unless Foo.new < 2
```

Can't run this MRI due to the JS, but here is the actual results from opal-0-9-stable:
```
no class no
class w/o parens yes
class w/ parens no
guard w/ parens
```

The result should be:
```
no class no
class w/o parens no
class w/ parens no
guard w/ parens
guard w/o parens
```

This problem could manifest itself anywhere the `nil` JS value is used in expressions. The first case I encountered was that it breaks module comparison unless parentheses are used.

**Comment by wied03** [comment]
Causing https://github.com/opal/opal-rspec/issues/51

**Comment by wied03** [comment]
Simplifying this has the same result:

Test Source:

```ruby
module Mod2
end

class Bar
end

def report(klass)
  puts "#{klass} < mod2 ? #{klass < Mod2 ? 'yes' : 'no'}"
end

report Bar
```

MRI:
```
Bar < mod2 ? no
```

Opal 0.9 final:
```
Bar < mod2 ? yes
```

**Comment by wied03** [comment]
Rubyspecs should be catching this with existing tests but the behavior of this code when I add it into ```rubyspec/core/module/lt_spec.rb``` is different than when I run the standalone code above.

**Comment by iliabylich** [comment]
It doesn't look like https://github.com/opal/opal/commit/cbb05e53de526cd234e18a4ce5f095b9e27d129d causes it. And it also happens for a regular `if`, not just with a ternary operator.
``` ruby
value = Foo.new

if value < 2
  puts 'yes'
else
  puts 'no'
end
```
Compiles to:
``` js
value = $scope.get('Foo').$new();
  if ($rb_lt(value, 2)) {
    return self.$puts("yes")
    } else {
    return self.$puts("no")
  };
```

Where `$rb_lt` returns `nil` (which is `true` in js). But usually when there's no operators in the `if` clause, it compiles
``` ruby
if value
  puts 'yes'
else
  puts 'no'
end
```
To:
``` js
if ((($a = self.$value()) !== nil && (!$a.$$is_boolean || $a == true))) {
    return self.$puts("yes")
    } else {
    return self.$puts("no")
  }
```

The issue seems to happen [here](https://github.com/opal/opal/blob/master/lib/opal/nodes/helpers.rb#L119). Expression `sexp` here is `(:call, (:call, nil, :value, (:arglist)), :<, (:arglist, (:int, 2)))` and its type = `:call` and the constant `Compiler::COMPARE` includes all comparison symbols (`["<", ">", "<=", ">="]`), so opal doesn't wrap it with a regular `tmp = (condition) && tmp !== nil && tmp == true)`.

Last time this file was changed more than a year ago :smile: (So opal 0.8.0 also [has](http://opalrb.com/try/?code:class%20Foo%0A%20%20def%20%3C(*)%0A%20%20%20%20nil%0A%20%20end%0Aend%0A%0Avalue%20%3D%20Foo.new%0A%0Aif%20value%20%3C%202%0A%20%20puts%20%27%3C%27%0Aelse%0A%20%20puts%20%27%3E%27%0Aend%0A%0Aputs%20RUBY_ENGINE_VERSION) this issue)

This code of the compiler looks incorrect. @elia Do you know why it was made this way? Objects comparison can be overwritten, so expression with any comparison definitely should be wrapped.

**Comment by iliabylich** [comment]
@elia As I understand, `js_truthy_optimize` should optimize only internal method stuff, like `block_given?` or `self` (that can't be overwritten and always returns non-nil object)

**Comment by wied03** [comment]
Didn't say it causes it. Said it revealed it because the module code is
using nil now. :)

**Comment by elia** [comment]
This looks like the same issue of #1270, ternary and `if` are both parsed as `if`s, didn't have the time to dig deeper into this but the problem I faced in #1270 is that the return is placed on individual branches and there are some combinations for which the the `if` is wrapped in a `function` but is missing a `return`.

**Comment by wied03** [comment]
I would say let's back out the JS `nil` usage in these hot spots that is breaking big stuff. Given the increasing usage of `nil` in JS code within corelib though, the root cause (that you guys are pointing out) probably needs to be fixed.

**Comment by wied03** [comment]
I'm taking a crack at it, will submit a PR for review

**Comment by wied03** [comment]
Take a look @ the PR and see what you think.

**Comment by wied03** [comment]
And thanks @iliabylich for pointing me to the right place to fix.

**Comment by wied03** [comment]
This is fixed in opal 0.9 stable and will be in master shortly
